ValveResourceFormat/Resource/ResourceTypes/KV3.cs

```diff
@@ -0,0 +1,592 @@
+                } else {
+                    objStack.Peek().AddProperty(currentName, new KVValue(typeof(Int32), Int32.Parse(currentString.ToString(), CultureInfo.InvariantCulture)));
+                }
+                return;
+            }
+
+            //Stop reading the number once whtiespace is encountered
+            if (whiteSpace.Contains(c))
+            {
+                stateStack.Pop();
+                //Distinguish between doubles and ints
+                if (currentString.ToString().Contains('.'))
+                {
+                    objStack.Peek().AddProperty(currentName, new KVValue(typeof(Double), Double.Parse(currentString.ToString(), CultureInfo.InvariantCulture)));
+                }
+                else
+                {
+                    objStack.Peek().AddProperty(currentName, new KVValue(typeof(Int32), Int32.Parse(currentString.ToString(), CultureInfo.InvariantCulture)));
+                }
+                return;
+            }
+
+            currentString.Append(c);
+        }
+
+        //Read an array
+        private static void ReadValueArray(char c)
+        {
+            //This shouldn't happen
+            if (!whiteSpace.Contains(c) && c != ',')
+            {
+                Console.WriteLine("ERROR");
+            }
+
+            //Just jump to seek_value state
+            stateStack.Push(State.SEEK_VALUE);
+        }
+
+        //Read a flagged value
+        private static void ReadValueFlagged(char c)
+        {
+            //End at whitespace
+            if (whiteSpace.Contains(c))
+            {
+                stateStack.Pop();
+                objStack.Peek().AddProperty(currentName, new KVValue(typeof(FlaggedString), currentString.ToString()));
+                return;
+            }
+
+            currentString.Append(c);
+        }
+
+        //Read comments
+        private static void ReadComment(char c)
+        {
+            //Check for multiline comments
+            if (currentString.Length == 1 && c == '*')
+            {
+                stateStack.Pop();
+                stateStack.Push(State.COMMENT_BLOCK);
+            }
+
+            //Check for the end of a comment
+            if (c == '\n') {
+                stateStack.Pop();
+                return;
+            }
+
+            if (c != '\r')
+            {
+                currentString.Append(c);
+            }
+        }
+
+        //Read a comment block
+        private static void ReadCommentBlock(char c)
+        {
+            //Look for the end of the comment block
+            if (c == '/' && currentString.ToString().Last() == '*')
+            {
+                stateStack.Pop();
+            }
+
+            currentString.Append(c);
+        }
+    }
+
+    //Data structure to contain the file contents
+    public class KVFile
+    {
+        public String Header;
```

Comment:
Why are these public? Consider making this `readonly`, with a public getter if necessary.